[WIP] Make deviator() and second_invariant() consistent with plane strain assumption in 2D

YiminJin opened the pull request:
This PR modifies the functions `deviator()` and `second_invariant()` to make them consistent with the plane strain assumption in 2D. The modified functions are placed in namespace `Utilities::Tensors`. For details of the mathematics, please refer to #6434 and #6459. A direct impact of this modification is that it sharpen the shear bands in plastic models with nonzero dilation angle.

This PR is a follow-up of #6373, but it will affect other material models. Also, we need several test problems to show the plane strain functions work well (or to find out if we can do better), which have not been implemented. Please make comments if you have suggestions or questions about this PR.

bangerth commented on tests/spiegelman_fail_test.cc:
```diff
@@ -204,7 +204,7 @@ namespace aspect
-    compute_second_invariant(const SymmetricTensor<2,dim> strain_rate, const double min_strain_rate) const
+    compute_Utilities::Tensors::deviatoric_tensor_inv2(const SymmetricTensor<2,dim> strain_rate, const double min_strain_rate) const
     {
       const double edot_ii_strict = std::sqrt(strain_rate*strain_rate);
```
It's worth mentioning that @cedrict and I looked into this at some point in the past and found that people have all sorts of incompatible definitions of the second invariant. Sometimes it included a factor of 2, sometime it didn't, and similar shenanigans. I would suggest sticking closely to the definition we had previously used, and only change 1/dim to 1/3.

YiminJin replied:
Sorry, this is my miss.

YiminJin commented on tests/spiegelman_fail_test.cc:
```diff
@@ -86,7 +86,7 @@ namespace aspect
     {
       public:
-        double compute_second_invariant(const SymmetricTensor<2,dim> strain_rate, const double min_strain_rate) const;
+        double compute_Utilities::Tensors::deviatoric_tensor_inv2(const SymmetricTensor<2,dim> strain_rate, const double min_strain_rate) const;
```
No, it is my mistake. I just used a script to replace the string `second_invariant` by `Utilities::Tensors::deviatoric_tensor_inv2` in all the files.

Do you think the second invariant of strain rate tensor should be changed here? I do not know if Spiegelman use a special function here (the function calculates the norm of the strain rate tensor).

YiminJin commented on include/aspect/utilities.h:
```diff
@@ -1407,6 +1407,29 @@ namespace aspect
+       */
+      template <int dim>
+      double
+      deviatoric_tensor_inv2(const SymmetricTensor<2,dim> &input);
```
I do not know. I agree that `plane_strain` should be added to the name, but `deviatoric` should also be added, because the function returns the correct result only when the input tensor is deviatoric. Yet, we are not able to check if the input tensor is deviatoric, for the deviator of 2D plane strain tensor is not "deviatoric" in the common sense...

gassmoeller commented on source/simulator/assemblers/newton_stokes.cc:
```diff
@@ -448,7 +448,7 @@ namespace aspect
 
               if (enable_elasticity)
-                data.local_rhs(i) += ( deviator(elastic_out->elastic_force[q])
+                data.local_rhs(i) += ( elastic_out->elastic_force[q]
```
why is this no longer the deviatoric force?

YiminJin replied:
Actually I do not know. But the Stokes also uses the full elastic force, and the deviatoric elastic force does not produce sharp shear bands in the strip footing test. I need some time to figure out why.

gassmoeller reviewed (commented):
Just a few small comments I saw while reading through the PR.

YiminJin commented on source/material_model/rheology/elasticity.cc:
```diff
@@ -858,7 +858,7 @@ namespace aspect
 
         const SymmetricTensor<2, dim>
-        edot_deviator = deviator(strain_rate) + 0.5 * stress_0_advected / elastic_viscosity
+        edot_deviator = strain_rate + 0.5 * stress_0_advected / elastic_viscosity
```
Yes. I think we need to be more careful when using `consistent_deviator()`, since we cannot apply it more than once to a symmetric tensor. I move the deviator here to the place where it is called --- `MaterialModel::Rheology::ViscoPlastic::compute_isostrain_viscosities` (`source/material_model/rheology/visco_plastic.cc, line 310). But I need to think this over, because under this change the Newton assembler gets the full strain rate instead of deviatoric strain rate.

bangerth replied:
By the way, is the name of the variable now wrong?

YiminJin commented on source/utilities.cc:
```diff
@@ -3325,6 +3325,33 @@ namespace aspect
+        return output;
+      }
+
+
```
Sure. I also add another empty line between `to_voigt_stiffness_vector` and `levi_civita` to make the format consistent.

YiminJin commented:
@gassmoeller The reasons for removing the `deviator` when calculating the viscoelastic strain rate and assembling the system rhs are:
1. When the material model is incompressible, the top-left block of the Stokes system should be
$\int_\Omega2\eta\nabla_S\varphi : \varepsilon d\Omega$, (1)
not
$\int_\Omega\nabla_S\boldsymbol\varphi : dev(\varepsilon) d\Omega$, (2)
because if we use (2), then the matrix becomes
$\int_\Omega2\eta\nabla_S\varphi : \left[\nabla_S\varphi - \frac{1}{3}(\nabla\cdot\varphi)I\right] d\Omega$,
which is not symmetric. Therefore, when we use Newton method, the system Jacobian should be
$\int_\Omega(2\eta\nabla_S\varphi : \varepsilon + 2\nabla_S\varphi : \varepsilon\otimes\frac{\partial\eta}{\partial\varepsilon} : \nabla_S\varphi)d\Omega$,
where $\varepsilon$ is the full strain rate.

I know it contradicts my modifications on the Newton assembler last year, but I cannot remember why I use deviatoric strain rate at that time...

2. In `source/material_model/rheology/elasticity.cc`, `deviator` is never applied to the stress tensor, for it is deviatoric itself. Although there would be numerical errors if we use field method to advect the deviatoric stress fields, I think it is better to keep the consistency between the codes material model and assembler.

The comments above are only simple analyses. I think our final choice should depend on experiments. I am working on it.

bangerth commented:
You already found out that
$\int_\Omega\nabla_S\varphi : (\nabla_S\varphi - \frac{1}{3}\nabla\cdot\varphi I)d\Omega 
 = \int_\Omega[\nabla_S\varphi : \nabla_S\varphi - \frac{1}{3}(\nabla\cdot\varphi)(\nabla\cdot\varphi)]d\Omega$

Moreover, we have
$\int_\Omega(\nabla\cdot\varphi I) : (\nabla_S\varphi - \frac{1}{3}\nabla\cdot\varphi I)d\Omega 
 = \int_\Omega[ (\nabla\cdot\varphi I) (I : \nabla_S\varphi - \frac{1}{3}\nabla\cdot\varphi I : I) ]d\Omega
 = \int_\Omega[ (\nabla\cdot\varphi I) (\nabla\cdot\varphi - \nabla\cdot\varphi) ]d\Omega
 = 0$
and as a consequence, 
$\int_\Omega\nabla_S\boldsymbol\varphi : dev(\varepsilon) d\Omega = \int_\Omegadev(varepsilon) : dev(\varepsilon) d\Omega$.

Both lines of the argument indicate that the system matrix is symmetric.

YiminJin commented:
I am trying out the benchmark in Spiegelman et al. (2016) with the modifications in this PR and #6546 . As a first step, I use the visco plastic model and the same material parameters as in `benchmarks/newton_newton_solver_benchmark_set/spiegelman_et_al_2016/input.prm` (except for the "regularization" viscosity). The prm file is:
[spiegelman_et_al_2016.prm.txt](https://github.com/user-attachments/files/20958453/spiegelman_et_al_2016.prm.txt)
The results:
![spiegelman](https://github.com/user-attachments/assets/93817313-4341-4c30-af9d-c4c934b3eb1e)
where $\eta^{vp}$ is the viscosity of the viscous damper. In all these tests, I cannot make the Newton solver converge to $10^{-4}$ within 100 iterations, and the Newton solver does not converge when `Use Newton residual scaling method` is off. This is apparently an unacceptable result. However, when I run the original benchmark with the main branch, the result looks like
![original](https://github.com/user-attachments/assets/7ce28035-2dcf-4963-bc23-0d49dbc44eba)
which is quite different from the figures in Spiegelman et al. (2016). Do we have model settings that can reproduce the results in the paper with a good convergence rate? @gassmoeller @MFraters

MFraters commented:
hmm, that means they have been broken somewhere along the way. Just checking, have you tried running the original files and material model here: https://github.com/geodynamics/aspect/tree/main/benchmarks/newton_solver_benchmark_set/spiegelman_et_al_2016?

YiminJin commented:
> Just checking, have you tried running the original files and material model here: https://github.com/geodynamics/aspect/tree/main/benchmarks/newton_solver_benchmark_set/spiegelman_et_al_2016

Yes. I only changed the refinement level from 4 to 6.

MFraters commented on source/utilities.cc:
```diff
@@ -3287,6 +3288,35 @@ namespace aspect
+      double consistent_second_invariant_of_deviatoric_tensor(const SymmetricTensor<2,dim> &t)
+      {
+        if (dim == 2)
+          return -( Utilities::fixed_power<2,double>(t[0][0])             // t11^2
```
why not simply `t[0][0]*t[0][0]`?

Same in other places.

MFraters reviewed (commented):
Thanks for doing all this work! That is great.

Do I understand it correctly that the convergence plot you show should be the same as the bottom left one in figure 5?

<img width="1216" height="788" alt="image" src="https://github.com/user-attachments/assets/11466236-1e80-4e48-b58c-f65b58df5e08" />

The stabilization definitely looks better, even though, looking at the Picard iteration the problems is more difficult, since that one is not converging. 

Do you have the same plots for the more difficult cases? It would be interesting to see how it behaves there.

YiminJin commented:
@MFraters The convergence behaviors for more difficult cases are shown below:
<img width="1705" height="529" alt="convergence_behavior" src="https://github.com/user-attachments/assets/dde7f379-9062-4497-b30a-42e39d950239" />

It can be seen that the convergence behavior of the current code is better indeed: the unstabilized Newton solver converges for the difficult cases with second-order rate. However, without stabilization the linear solver takes more iterations to converge, and in the most difficult case (Vel = 12.5 mm/y) the cheap iterations are even exhausted. So it is still worth using the stabilization in larger models.

YiminJin commented:
Supplement to my comments on my derivation of the analytical formula of system Jacobian:

The derivation is under plane-strain assumption, but the conclusion also applies to the original 2D formulation. In the original 2D formulation, we have
$(2\Vert\boldsymbol\varepsilon\Vert\Vert\boldsymbol\varepsilon'\Vert)^2 - (\Vert\boldsymbol\varepsilon\Vert^2 + \boldsymbol\varepsilon:\boldsymbol\varepsilon')^2 = -\dfrac{1}{4}(\varepsilon_{xx} + \varepsilon_{yy})^4$.
Hence, $\mathbb{E}$ still has a negative eigenvalue if we use full strain rate in the assembler, and the convergence rate is slowed down by PD stabilization.
However, I tried to use deviatoric strain rate in the Newton Stokes assembler in this branch (under plane strain assumption), and the resulting convergence behavior is a little bit worse than the full strain rate (the difference is small). Currently I do not understand why.

MFraters reviewed (commented):
Generally looks good to me. A few small comments to clear things up. 

I don't see a significant difference in the unstabilized convergence when it converges (maybe I am missing something), but the stabilized one is definitely better. 

Can you fix the tests? Than we can also see what difference it makes there.

@bangerth, can you have a look at the derivations?

YiminJin commented on benchmarks/newton_solver_benchmark_set/spiegelman_et_al_2016/drucker_prager_compositions.cc:
```diff
@@ -316,15 +319,11 @@ namespace aspect
                               const double regularization_adjustment = (ref_visc * ref_visc)
-                                                                       / (ref_visc * ref_visc + 2.0 * ref_visc * drucker_prager_viscosity
-                                                                          + drucker_prager_viscosity * drucker_prager_viscosity);
+                                                                       / Utilities::fixed_power<2,double>(ref_visc + drucker_prager_viscosity);
```
I did not read the source code of `Utilities::fixed_power()` when I made the change. Now I think you are right: using the function may not be faster. I have changed it to explicit form in a new commit.

YiminJin commented on doc/modules/changes/20250617_yiminjin:
```diff
@@ -0,0 +1,5 @@
+Changed: The deviator and second invariant of symmetric tensors are modified to
+be consistent with the plane strain assumption in 2D. It changes the outputs of
+compressible material models that are dependent on the deviatoric strain rate.
```
I forget why I wrote this. I deleted the second sentence in a new commit.

YiminJin commented:
Sorry, I made a big mistake in the last comment: I forgot that the 2D tensors in the channel flow benchmark have not been changed to plane-strain tensors.

When I tried to do the modification, I found something that looks weird: line 183 and line 189 of `benchmarks/newton_solver_benchmark_set/nonlinear_channel_flow/simple_nonlinear.cc` calculate the viscosity and the viscosity derivatives according to the first two equations in Appendix B1 of the original paper:
$\eta = \eta_0^{-1/n}\varepsilon_{II}^{1/n-1}$,
$\dfrac{\partial\eta}{\partial\boldsymbol\varepsilon} = \eta\bigg(\dfrac{1}{n}-1\bigg)\dfrac{\boldsymbol\varepsilon}{\Vert\boldsymbol\varepsilon\Vert^2}$
However, line 180 multiplies `edot_ii` by 2, which leads to
$\eta = \eta_0^{-1/n}(2\varepsilon_{II})^{1/n-1}$
Is there something I am missing? @MFraters

YiminJin commented:
I also made a mistake when computing the deviatoric strain rates in system Jacobian: I forgot to change `deviator()` to `Utilities::Tensors::consistent_deviator()`. Here are the correct results for DSR in Spiegelman et al. (2016) benchmark:
<img width="1600" height="1050" alt="figure_4" src="https://github.com/user-attachments/assets/52c3625d-3fc8-417d-b369-e47a5891bea4" />
It is almost the same as the full strain rate case. I also compared the distribution of SPD factor, and the results are again the same. I think this is because the material model is incompressible, hence $\dfrac{\eta}{\boldsymbol a : \boldsymbol b}$ is very close to -1, which leads to
$\alpha_{SPD} = -c_{safety}\dfrac{\eta}{\boldsymbol a : \boldsymbol b}\approx c_{safety}$
in the yielding regions. 

@gassmoeller Do you remember how we came to the conclusion that using deviatoric strain rate when stabilized is faster than using full strain rate?

gassmoeller commented:
@YiminJin Concerning the deviatoric strain rate: I think the relevant PR where I tested this was https://github.com/geodynamics/aspect/pull/6159#issuecomment-2503528026, and in particular this code: https://github.com/geodynamics/aspect/pull/6159/files#diff-67a939b674da49b34479b331895052a08d8712da7befddadaba9390cd782ee05R183. As far as I can tell we did not decide to use the deviatoric strain rate for the stabilized version because it is faster, it seems like there was a theoretical argument that the stabilization factor was only safe if applied to a deviatoric strain rate. In other words the guarantee that the stabilization makes the matrix positive definite is only valid if the used strain rate is deviatoric. But I have to say I dont remember the details or if I ever made that derivation myself (I would guess I accepted it from the paper or some earlier PR discussion). From looking at the earlier PR my main point was that we need to use the *full* strain rate for the *unstabilized* version in order to guarantee the expected convergence rate of the Newton method. If your version here works with the full strain rate for both versions so much the better.

YiminJin commented:
I plot the number of cheap linear iterations in each Newton iteration for the Spiegelman benchmark:

1. Use __full__ strain rate in `NewtonStokesIncompressibleTerms`:

  a. residuals:
<img width="1600" height="1050" alt="figure_4" src="https://github.com/user-attachments/assets/c650f2ff-950e-4ef1-9e13-7d097d37af21" />

  b. number of cheap iterations:
<img width="1941" height="1439" alt="convergence_behavior" src="https://github.com/user-attachments/assets/65ca6f9c-3ee3-433a-8ba6-91ac224eb026" />


2. Use __deviatoric__ strain rate in `NewtonStokesIncompressibleTerms`:

  a. residuals:
<img width="1600" height="1050" alt="figure_4" src="https://github.com/user-attachments/assets/1286eb41-92d5-4347-a12e-8354ca9023b8" />

  b. number of cheap iterations:
<img width="1941" height="1439" alt="convergence_behavior" src="https://github.com/user-attachments/assets/7803b931-085a-40eb-b8c9-8547e18213c7" />

Remark: 
1. The cheap iterations are limited by 1000;
2. Five cases report linear solver errors (The iterative (top left) solver did not converge): 
     Vel = 12.5 mm/yr, eta_ref = 5e24, mLT = 1e-8, 0 Picard, unstabilized, with __full__ strain rate;
     Vel = 12.5 mm/yr, eta_ref = 5e24, mLT = 1e-8, 5 Picard, unstabilized, with __full__ strain rate;
     Vel = 12.5 mm/yr, eta_ref = 5e24, mLT = 1e-8, 25 Picard, unstabilized, with __full__ strain rate;
     Vel = 12.5 mm/yr, eta_ref = 5e24, mLT = 1e-8, 0 Picard, unstabilized, with __deviatoric__ strain rate;
     Vel = 12.5 mm/yr, eta_ref = 5e24, mLT = 1e-8, 5 Picard, unstabilized, with __deviatoric__ strain rate;

It can be seen that:
1. When not stabilized, using __full__ strain rate generally takes fewer Newton iterations than using __deviatoric__ strain rate, but the convergence is more difficult (takes more cheap linear iterations, and easier to fail);
6. When stabilized, the convergence behaviors of __full__ strain rate and __deviatoric__ strain rate are similar.

Fact 1 confirms the point that using __full__ strain rate breaks the symmetry and semi-positive-definiteness of the top-left block. In the easiest case (Vel = 2.5 mm/yr, eta_ref = 1e23), the convergence curve of stabilized and unstabilized cases are almost the same when using __deviatoric__ strain rate, which suggests that using __deviatoric__ strain rate is equivalent to symmetrizing the matrix (changing the descending direction, slowing down the convergence, but restoring the symmetry and makes the linear solver stabler).

The reason that using deviatoric strain rate or not affects the convergence behavior when using __incompressible__ model is unclear. I guess it comes from numerical error. The distribution of strain rate and velocity divergence in case {Vel = 5 mm/yr, eta_ref = 1e24, mLT = 1e-8} are:
<img width="1780" height="393" alt="numerical_error" src="https://github.com/user-attachments/assets/aee59cfd-4ac0-4579-914e-988c24e00b90" />

It can be seen that the numerical error (velocity divergence) is in the same magnitude as the strain rate. So the absolute value of the __negative__ eigenvalue of the tangent operator when using __full__ strain rate is large at some nodes.

YiminJin commented:
I tested the strip footing model:
<img width="981" height="689" alt="strip_footing" src="https://github.com/user-attachments/assets/cd1aeafa-71bd-4683-91d1-60cfd50b6616" />

The Newton solver parameters are set as follows:
| Parameter | Value |
| --- | --- |
| Max nonlinear iterations | 100 |
| Nonlinear solver tolerance | 1e-5 |
| Max pre-Newton nonlinear iterations | 3 |
| Max Newton line search iterations | 3 |
| Maximum linear Stokes solver tolerance | 1e-2 |
| Use Newton residual scaling method | false |
| Stabilization preconditioner | SPD |
| Stabilization velocity block | SPD |

The convergence behaviors of the __full__ strain rate case and the __deviatoric__ strain rate case are as follows:
<img width="1495" height="776" alt="convergence_behavior" src="https://github.com/user-attachments/assets/f47296e0-1326-489f-8e86-ef4dad16bf1c" />

Clearly, the __full__ strain rate prevails. It can be explained by the fact that in compressible/prescribed-dilation models, the assembler `NewtonStokesCompressibleStrainRateTerms` has already removed the volumetric part, so if we use __deviatoric__ strain rate in `NewtonStokesIncompressibleTerms`, then we actually minus the volumetric part __twice__, which does not matter in 2D or 3D cases, but is incorrect in plane-strain cases.

gassmoeller commented on doc/modules/changes/20250617_yiminjin:
```diff
@@ -0,0 +1,4 @@
+Changed: The deviator and second invariant of symmetric tensors are modified to
+be consistent with the plane strain assumption in 2D.
```
Can you add a sentence about which models will be affected? E.g.:

```suggestion
Fixed: The deviator and second invariant of symmetric tensors were modified to
be consistent with the plane strain assumption in 2D. This fixes a bug
in the strain rate and strain rate invariant computation of many material
models for compressible 2D models. It also results in better
nonlinear solver convergence due to some fixes to the Newton solver.
```

gassmoeller reviewed (commented):
Thanks for putting all the effort into this! I will take another look at the main changes again, but here are a few small comments for now.

YiminJin commented on source/simulator/solver/stokes_matrix_free_local_smoothing.cc:
```diff
@@ -564,7 +564,7 @@ namespace aspect
                             effective_strain_rate = elastic_out->viscoelastic_strain_rate[q];
                           else if ((this->get_newton_handler().parameters.velocity_block_stabilization & Newton::Parameters::Stabilization::PD) != Newton::Parameters::Stabilization::none)
-                            effective_strain_rate = deviator(effective_strain_rate);
+                            effective_strain_rate = Utilities::Tensors::consistent_deviator(effective_strain_rate);
```
Thanks for finding the flaws. The variable name has been fixed.

The GMG solver requires more attention. Besides changing deviatoric strain rate to full strain rate, I found two other problems:
1. The Newton derivatives related to the volumetric part is missing;
2. When material averaging is set to none, the Newton scheme does not work because the uninitialized `viscosity_derivative_averaging_weights` are assembled into the system.

I fixed the problems in a new commit. However, I cannot make the strip footing model run with GMG solver. The Kaus benchmark (with $\phi=\psi=30^\circ$) works, but it also requires Newton residual scaling method to achieve convergence. Below are the convergence behaviors with and without the Newton derivatives related to the volumetric part (only nonlinear iterations 10 to 19 are presented):
<img width="1495" height="776" alt="convergence_behavior" src="https://github.com/user-attachments/assets/03fdd2e7-9249-437e-867d-9a53f22e3513" />

The prm file is:
[kaus_2010_extension.prm.txt](https://github.com/user-attachments/files/23538072/kaus_2010_extension.prm.txt)

It can be seen that adding the volumetric part does not lead to a big improvement in convergence rate (except for the 18th nonlinear iteration). Nevertheless, I feel it is safer to follow the ``correct'' formula.

Please let me know if there are problems in my codes, or more benchmarks to run.

YiminJin commented on source/simulator/solver/matrix_free_operators.cc:
Thanks for finding this! It is a bug indeed, and it comes from a trick I used to reduce the computation. Let me break it down:

The top-left block of the system Jacobian w.r.t. incompressible terms can be expressed as
$$2\sum_qw_qJ_q\nabla_s\boldsymbol\varphi_{Iq} : \boldsymbol\varepsilon_q\otimes\dfrac{\partial\eta_q}{\partial\boldsymbol\varepsilon_q} : \nabla_s\boldsymbol\varphi_{Jq}$$
When using cellwise averaging, the expression becomes
$$2\sum_qw_qJ_q\nabla_s\boldsymbol\varphi_{Iq} : \boldsymbol\varepsilon_q\otimes\sum_r\tilde w_r\dfrac{\partial\eta_r}{\partial\boldsymbol\varepsilon_r} : \nabla_s\boldsymbol\varphi_{Jr}$$
where $\tilde w_r$ is the averaging weight. Now do the symmetrizing:
$$\sum_qw_qJ_q\nabla_s\boldsymbol\varphi_{Iq} : \boldsymbol\varepsilon_q\otimes\sum_rw_r\dfrac{\partial\eta_r}{\partial\boldsymbol\varepsilon_r} : \nabla_s\boldsymbol\varphi_{Jr} + \sum_r\tilde w_r\nabla_s\boldsymbol\varphi_{Ir} : \dfrac{\partial\eta_r}{\partial\boldsymbol\varepsilon_r}\otimes\sum_qw_qJ_q\boldsymbol\varepsilon_q : \nabla_s\boldsymbol\varphi_{Jq}$$
Let us focus on the second term. It can be rewritten as
$$\sum_q\sum_rw_qJ_q\tilde w_r\nabla_s\boldsymbol\varphi_{Ir} : \dfrac{\partial\eta_r}{\partial\boldsymbol\varepsilon_r}\otimes\boldsymbol\varepsilon_q : \nabla_s\boldsymbol\varphi_{Jq}$$
I thought that since we sum over $q$ and $r$, we can swap the two subscripts:
$$\sum_q\sum_r\color{red}w_qJ_q\color{black}\tilde w_q\nabla_s\boldsymbol\varphi_{Iq} : \dfrac{\partial\eta_q}{\partial\boldsymbol\varepsilon_q}\otimes\boldsymbol\varepsilon_r : \nabla_s\boldsymbol\varphi_{Jr}$$
This is the current implementation. However, **I forgot that the subscripts of $w_qJ_q$ should also change**. In the matrix-free framework, there is no way to do this. So the trick is wrong.
Moreover, **I forgot this issue for the compressible terms**. I think these are the reasons for the GMG solver to be unstable right now.
I will try your suggestions to fix this problem.

YiminJin replied:
After further study I found that the current implementation is not just a "trick", but the only way I found to do the symmetrization. 

First, we cannot pre-compute the averaged values in `StokesMatrixFreeHandlerLocalSmoothingImplementation::evaluate_material_model()`, because the quantity to be averaged is the product of $\partial\eta/\partial\boldsymbol\varepsilon$ and $\nabla_s\boldsymbol\varphi_J$, and in the matrix-free framework $\boldsymbol\varphi_J$ is bound to the solution vector (which changes after each linear iteration). Thus, we must do the averaging in `local_apply()` and try to separate $w_qJ_q\nabla_s\boldsymbol\varphi_{Iq}$ from the rest terms. 

Currently, the only way I find is to rewrite the second term by
$\sum_q\sum_r w_qJ_q \dfrac{w_rJ_r}{w_qJ_q}\tilde w_q\nabla_s\boldsymbol\varphi_{Iq} : \dfrac{\partial\eta_q}{\partial\boldsymbol\varepsilon_q}\otimes\boldsymbol\varepsilon_r : \nabla_s\boldsymbol\varphi_{Jr}$
and the symmetric tensor to submit (to be multiplied by $w_qJ_q\nabla_s\boldsymbol\varphi_{Iq}$) for each quadrature point q is
$\tilde w_q\dfrac{\partial\eta_q}{\partial\boldsymbol\varepsilon_q}\dfrac{\sum_rw_rJ_r\boldsymbol\varepsilon_r : \nabla_s\boldsymbol\varphi_{Jr}}{w_qJ_q}$
Similarly, the term corresponding to compressibility is
$-\dfrac{\tilde w_q}{3}\dfrac{\partial\eta_q}{\partial\boldsymbol\varepsilon_q}\dfrac{\sum_rw_rJ_r(\nabla\cdot\boldsymbol u_r)(\nabla\cdot\boldsymbol\varphi_{Jr})}{w_qJ_q}$

I tried this, but the convergence behavior is not improved. I still cannot run the strip footing model even with Newton residual scaling method.

Are there mistakes in my derivation @bangerth @tjhei ?

The pdf below is the original documentation for the viscosity averaging issues in Newton solver:
[cell_averaging.pdf](https://github.com/user-attachments/files/23639485/cell_averaging.pdf)

bangerth commented on source/simulator/assemblers/newton_stokes.cc:
```diff
@@ -687,6 +822,58 @@ namespace aspect
+
+      std::vector<double> deta_deps_times_grads_phi_u(stokes_dofs_per_cell);
+      std::vector<double> deta_dp_times_phi_p(stokes_dofs_per_cell);
+      std::vector<double> div_u_times_div_phi_u(stokes_dofs_per_cell);
```
If I see this right, you only use these variables when you do averaging, right? If so, perhaps declare these variables here as empty arrays, and then resize them to the correct size inside the `if` condition immediately below -- this way, you don't pay the price of memory allocation if you don't actually need the variables.

YiminJin commented:
@bangerth @gassmoeller I shall apologize for being absent for a long time. I thought I could come back to this after finishing my own tasks during Christmas, but I felt so exhausted that I took a few days' break.

@bangerth Thank you for going through this PR. I have fixed the space-allocating problem in a new commit (there is a small problem. Please take a look at my reply to your comment). About the format problems: I have fixed the problems you pointed out, but I believe there are more problems (like extra white spaces) that cannot pass the format check. However, if I use astyle to reformat the file, there would be some annoying changes. For example, line 89-90 in `source/simulator/assemblers/newton_stokes.cc` is:
```
      const std::shared_ptr<const MaterialModel::MaterialModelDerivatives<dim>> derivatives
        = scratch.material_model_outputs.template get_additional_output_object<MaterialModel::MaterialModelDerivatives<dim>>();
```
After reformatting, it becomes
```
      const std::shared_ptr<const MaterialModel::MaterialModelDerivatives<dim>> derivatives
                                                                             = scratch.material_model_outputs.template get_additional_output_object<MaterialModel::MaterialModelDerivatives<dim>>();
```
I think it is because `>>` is identified as a in/out signal. Are there some better ways to reformat the modified files?

tjhei commented:
> I think it is because `>>` is identified as a in/out signal. Are there some better ways to reformat the modified files?

You could use ``> >`` and see if that helps.

YiminJin commented:
I have fixed most of the problems in this PR and cleaned up the commit history (only 4 commits are left). The failed test seems to be related to the problem inside the master branch of dealii (it runs on my machine, with dealii-9.6). There might be some hidden bugs left (for example, there might be some plugins in cookbook or benchmark that still use the inconsistent 2D deviator), but I think they won't harm the performance of the software, and we can let them be discovered by some sharp-eyed fellow (like the problem hidden in the nonlinear channel flow benchmark). Do you think this PR is prepared to be merged?@gassmoeller @bangerth @MFraters ?

gassmoeller reviewed (approved):
I think it is time to close this chapter and merge the PR. Thanks again @YiminJin for this massive amount of work, I think this significantly improves our material model consistency and the Newton solver performance! We appreciate the work you do!

As Wolfgang already mentioned: For the future please try to separate fixes into individual PRs. This PR could have been much less work for us and for you if we could have handled the material model change and the change to the Newton solver in two separate PRs. I know how tempting it is to include each fix you find into the current branch, but it complicates things during review.

Just to be sure I will restart the deal.II development tester, it should be fixed by now, and I just want to avoid to introduce the next problem into it. When all testers are done, this is ready to merge.

gassmoeller reviewed (commented):
Ok, sorry I found one more thing that needs to be fixed (the test change files should not be committed in this PR). 

And I would like to get confirmation on one more benchmark, because we so far only discussed the plasticity benchmarks and this PR also modifies the elasticity terms. Someone would need to check the viscoelastic bending beam benchmark and compare the main branch with this PR. (I will see if I can get to that soon, but if someone else want to take the lead, feel free to).

gassmoeller commented:
Thanks for cleaning the files. 

No need to rerun the benchmark again, I have confirmed that the bending beam benchmark still rebounds as before (see figure below). The stress evolution has changed a bit (lower panel below, this is consistent with the changes in the test results as well). However, I think this is not too surprising given that you unified the elastic force term for the Newton solver to the way we used it in the Picard solver. The terms should definitely be the same for both solvers.

I think this settles it then, this PR is ready to merge. Thanks again for all your efforts!

<img width="3000" height="1800" alt="5_viscoelastic_bending_beam_dte_particles_isnot_dtc_depth_main_fieldprop_np_test_LLS_QLS" src="https://github.com/user-attachments/assets/5b0ec8e2-02df-4d63-ae4b-eb10b5cda9b2" />
